cpp/src/arrow/dataset/discovery.cc

```diff
@@ -134,8 +135,26 @@ Result<std::shared_ptr<DatasetFactory>> FileSystemDatasetFactory::Make(
 Result<std::shared_ptr<DatasetFactory>> FileSystemDatasetFactory::Make(
     std::shared_ptr<fs::FileSystem> filesystem, const std::vector<fs::FileInfo>& files,
     std::shared_ptr<FileFormat> format, FileSystemFactoryOptions options) {
+  // Discover files in directories and globs
+  std::vector<fs::FileInfo> discovered_files;
+  for (const auto& file : files) {
+    if (file.IsDirectory()) {
+      fs::FileSelector file_selector;
+      file_selector.base_dir = file.dir_name();
+      file_selector.recursive = true;
+      ARROW_ASSIGN_OR_RAISE(auto folder_files, filesystem->GetFileInfo(file_selector));
+      std::move(folder_files.begin(), folder_files.end(),
+                std::back_inserter(discovered_files));
+    } else if (file.IsGlob()) {
+      ARROW_ASSIGN_OR_RAISE(auto files, filesystem->GetFileInfoGlob(file));
+      std::move(files.begin(), files.end(), std::back_inserter(discovered_files));
+    } else if (file.IsFile()) {
+      discovered_files.emplace_back(file);
+    }
+  }
+
```

Comment:
So this is something of an abstract concern but I think we have a bit of ambiguity between input & output types.

I see `FileInfo` as an "output type".  It is something you get back from a filesystem that describes something that it went out and discovered or inspected.

In the meantime, PlatformFilename and FileSelector are "input" types.  They are provided by the user for the purpose of obtaining FileInfo objects.

As a result I'm not sure it makes sense to have `FileType::Glob` because I would interpret that as the filesystem discovering a glob inside the filesystem.  For example `GetFileInfo` would never return `FileType::Glob`.

Instead, can we do something slightly different:

Instead of modifying the `Make` overload that accepts `const std::vector<fs::FileInfo>&` can we add a new overload that takes `const std::vector<fs::FileSelector>&`?

Then, instead of adding `FileType::Glob` can we add an `is_glob` property to `FileSelector` (if this is true then `base_dir` is expected to be a glob pattern and `recursive` will be ignored).

We could also then change the rules a little on `FileSelector` so that `base_dir` is allowed to represent a single file.  Then we can treat the "path" case (user doesn't specify it as a file or a directory) as a selector also.  That would allow us to move all the `GetFileInfo` calls out of the Substrait consumer.

CC @pitrou for a second opinion as he has a lot of experience here.